microsetta_public_api/config.py

```diff
@@ -1,6 +1,226 @@
 import os
 import json
-from microsetta_public_api.exceptions import ConfigurationError
+from typing import Union
+import jsonschema
+
+from microsetta_public_api.backend import constructors
+
+
+def component_schema(properties=None,
+                     required=None,
+                     named_components: Union[dict, bool] = False,
+                     construct=False,
+                     config=False,
+                     components=False,
+                     additional_properties=False,
+                     ):
+    """
+    Helper for defining a schema composed of a hierarchy of components
+
+    Parameters
+    ----------
+    properties : dict, optional
+        A dict of specifically named sub-components with schemas.
+    required : list, optional
+        Allows you to specify required sub-components of this component.
+    named_components : bool or dict, optional
+        False if no sub-components with arbitrary names should be allowed.
+        True, otherwise. Can also use a dictionary to specify the expected
+        schema of the sub-components.
+    construct : bool, optional
+        Indicates wheterh `construct` should be a requiref field for this
+        component. Default: False
+    config : bool, optional
+        Indicates whether `config` should be a required field for this
+        component. Default: False
+    components : bool, optional
+        Indicates whether `components` should be a required field for this
+        component. Default: False
+    additional_properties : bool, optional
+        Indicates whether properties in addition to `construct`,
+        `config`, or `components` should be allowed. Default: False
+
+    Returns
+    -------
+    dict
+        The output schema
+
+    """
+    if properties is None:
+        properties = {}
+    required_parts = [element for element, is_required in
+                      zip(['construct', 'config', 'components'],
+                          [construct, config, components])
+                      if is_required
+                      ]
```

Comment:
🤯 
This is just 3 if statements that result in appending strings to a list right?  

Oh python, you make the simplest things look so difficult!

Reply:
Yup.

Reply:
I think this would be clearer if written as:

```python
required_parts = []
elements = ['construct', 'config', 'components']
required = [construct, config, components]
for element, is_required in zip(elements, required):
    if is_required:
        required_parts.append(element)
```

Reply:
Nice!

Reply:
```
construct and required_parts.append('construct')
config and required_parts.append('config')
components and required_parts.append('components')
```

😄

Reply:
> construct and required_parts.append('construct')
> config and required_parts.append('config')
> components and required_parts.append('components')

That's hilarious 😂

Reply:
Ya, that also works too. However, it can be surprising to someone not familiar with it as there is implicit behavior and boolean logic on against effectively null:

```python
>>> x = []
>>> type(x.append(10))
<class 'NoneType'>
```

Reply:
Agreed. I am about to push some updates that kind of scraps this logic anyways, so wouldn't worry about it too much.